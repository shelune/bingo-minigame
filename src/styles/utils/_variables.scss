$black: #202020;
$white: #ffffff;
$orange: #ff6600;
$orange-dark: darken($orange, 10%);
$gray: #656565;
