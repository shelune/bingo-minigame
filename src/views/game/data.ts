import { ComponentProps, useEffect, useState } from "react";
import differenceInSeconds from "date-fns/differenceInSeconds";

import { Tile } from "../../components/board/board";
import { GameView } from "./game";
import {
  checkColumn,
  checkDiagonal,
  checkRow,
  getBoardNumbers,
  getNumberFromRange,
} from "./utils";
import { useStorage } from "../../utils/hooks/useStorage";
import { PlayerData } from "../../components/profile/profile";

export type WinCondition = "column" | "row" | "diagonal" | "";

export const useData = (
  username: string
): Omit<ComponentProps<typeof GameView>, "username"> => {
  const getNewBoard = () => {
    const numbers = getBoardNumbers().flat();
    const tiles: Tile[] = numbers.map((number, idx) => {
      return {
        id: idx,
        value: number,
        marked: false,
      };
    });
    return tiles;
  };
  const [tiles, setTiles] = useState(getNewBoard());
  const [winCondition, setWinCondition] = useState<WinCondition>("");
  const [picked, setPicked] = useState<number[]>([]);
  const [timeStart, setTimeStart] = useState<Date | null>(null);
  const [elapsedTime, setElapsedTime] = useState(0);

  const { storedValue, setValue, clearKey } = useStorage<PlayerData>(
    `bingo-player-${username}`
  );

  const onReset = () => {
    // Save to profile if win
    if (winCondition) {
      if (!storedValue) {
        setValue({
          name: username,
          gamesCount: 1,
          totalTimeSpent: elapsedTime,
          recentGames: [
            {
              winCondition,
              turns: picked.length,
              elapsedTime: elapsedTime,
            },
          ],
        });
      } else {
        const { gamesCount, totalTimeSpent, recentGames } = storedValue;
        setValue({
          ...storedValue,
          gamesCount: gamesCount + 1,
          totalTimeSpent: totalTimeSpent + elapsedTime,
          recentGames: [
            ...recentGames,
            {
              winCondition,
              turns: picked.length,
              elapsedTime: elapsedTime,
            },
          ].slice(0, 5),
        });
      }
    }

    // Reset
    const newBoard = getNewBoard();
    setTiles(newBoard);
    setWinCondition("");
    setPicked([]);
    setTimeStart(null);
    setElapsedTime(0);
  };

  const onAdvanceTurn = () => {
    const newNumber = getNumberFromRange(tiles.length, picked);
    setPicked([...picked, newNumber]);
    onMark(tiles, newNumber);
  };

  const onMark = (tiles: Tile[], id: number) => {
    if (picked.includes(id)) {
      return;
    }
    const newTiles = tiles.slice();
    const newTile = newTiles.find((tile) => tile.id === id);
    if (newTile) {
      newTiles[id].marked = true;
    }

    setPicked([...picked, id]);
    setTiles(newTiles);
    const hasRowWinner = checkRow(newTiles, id);
    const hasColumnWinner = checkColumn(newTiles, id);
    const hasDiagonalWinner = checkDiagonal(newTiles);
    if (hasRowWinner) {
      setWinCondition("row");
    }
    if (hasColumnWinner) {
      setWinCondition("column");
    }
    if (hasDiagonalWinner) {
      setWinCondition("diagonal");
    }
  };

  const onDeleteProfile = () => {
    clearKey();
  };

  useEffect(() => {
    if (timeStart === null && picked.length !== 0) {
      setTimeStart(new Date());
    }
  }, [picked, timeStart]);

  useEffect(() => {
    if (!!winCondition && timeStart) {
      const elapsed = differenceInSeconds(new Date(), timeStart);
      setElapsedTime(elapsed);
    }
  }, [timeStart, winCondition]);

  return {
    tiles,
    onMark,
    winCondition,
    onAdvanceTurn,
    onReset,
    picked,
    elapsedTime,
    profile: storedValue,
    onDeleteProfile,
  };
};
