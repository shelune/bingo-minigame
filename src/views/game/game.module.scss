@import "../../styles/utils/index";

.interactions {
  max-width: 550px;
  margin: 24px auto;
  display: flex;
  gap: 8px;
  justify-content: center;
}

.button {
  @include text-body;
  @include rounded-4;
  cursor: pointer;
  background-color: $orange;
  color: $white;
  outline: none;
  border: none;
  height: 64px;
  line-height: 64px;
  min-width: 120px;

  &:disabled {
    cursor: not-allowed;
    background-color: $gray;
  }
}
