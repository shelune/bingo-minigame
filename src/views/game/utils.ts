import { Tile } from "../../components/board/board";
import shuffle from "lodash.shuffle";

export function range(min: number, max: number) {
  return Array.from({ length: max - min + 1 }, (_, idx) => min + idx);
}

export function getNumberFromRange(range: number, excluding: number[]): number {
  const number = Math.round(Math.random() * (range - 1));
  return excluding.includes(number)
    ? getNumberFromRange(range, excluding)
    : number;
}

export function transpose(matrix: number[][]) {
  return matrix[0].map((_, colIdx) =>
    matrix.map((_, rowIdx) => matrix[rowIdx][colIdx])
  );
}

export function getBoardNumbers() {
  const numbers = [0, 1, 2, 3, 4].map((columnIdx) => {
    const selections = shuffle(range(columnIdx * 15 + 1, (columnIdx + 1) * 15));
    return [1, 2, 3, 4, 5].map((idx) => selections[idx]);
  });
  // transpose so the bingo card follows column not row
  const transposed = transpose(numbers);
  return transposed;
}

export function generateBoardTiles(numbers: number[]) {
  const tiles: Tile[] = numbers.map((number, idx) => {
    return {
      id: idx,
      value: number,
      marked: false,
    };
  });
  return tiles;
}

export function checkRow(board: Tile[], idxOnBoard: number) {
  const rowIdx = Math.floor(idxOnBoard / 5);
  const markedCount = [0, 1, 2, 3, 4].reduce((accu, curr) => {
    const boardIdx = rowIdx * 5 + curr;
    if (board[boardIdx].marked) {
      return accu + 1;
    }
    return accu;
  }, 0);
  return markedCount === 5;
}

export function checkColumn(board: Tile[], idxOnBoard: number) {
  const columnIdx = idxOnBoard % 5;
  const markedCount = [0, 1, 2, 3, 4].reduce((accu, curr) => {
    const boardIdx = curr * 5 + columnIdx;
    if (board[boardIdx].marked) {
      return accu + 1;
    }
    return accu;
  }, 0);
  return markedCount === 5;
}

export function checkDiagonal(board: Tile[]) {
  const diagonalForwardTiles = board.filter((tile) => {
    return tile.id > 0 && tile.id % 4 === 0 && tile.id < 24;
  });
  const diagonalBackwardTiles = board.filter((tile) => {
    return tile.id % 6 === 0;
  });
  return (
    diagonalForwardTiles.every((tile) => tile.marked) ||
    diagonalBackwardTiles.every((tile) => tile.marked)
  );
}
