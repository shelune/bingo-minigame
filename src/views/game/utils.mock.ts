import { generateBoardTiles, getBoardNumbers } from "./utils";

const numbers = getBoardNumbers().flat();
const baseBoard = generateBoardTiles(numbers);

export const mockBareBoard = [
  { id: 0, marked: false, value: 2 },
  { id: 1, marked: false, value: 16 },
  { id: 2, marked: false, value: 39 },
  { id: 3, marked: false, value: 55 },
  { id: 4, marked: false, value: 61 },
  { id: 5, marked: false, value: 9 },
  { id: 6, marked: false, value: 25 },
  { id: 7, marked: false, value: 32 },
  { id: 8, marked: false, value: 53 },
  { id: 9, marked: false, value: 67 },
  { id: 10, marked: false, value: 10 },
  { id: 11, marked: false, value: 19 },
  { id: 12, marked: false, value: 45 },
  { id: 13, marked: false, value: 59 },
  { id: 14, marked: false, value: 63 },
  { id: 15, marked: false, value: 13 },
  { id: 16, marked: false, value: 27 },
  { id: 17, marked: false, value: 43 },
  { id: 18, marked: false, value: 47 },
  { id: 19, marked: false, value: 64 },
  { id: 20, marked: false, value: 7 },
  { id: 21, marked: false, value: 18 },
  { id: 22, marked: false, value: 38 },
  { id: 23, marked: false, value: 58 },
  { id: 24, marked: false, value: 71 },
];

export const mockWonBoardColumn = baseBoard.map((tile) => {
  // board win by 2nd column
  if (tile.id % 5 === 1) {
    return {
      ...tile,
      marked: true,
    };
  }
  return tile;
});

export const mockWonBoardRow = baseBoard.map((tile) => {
  // board win by first row
  if (Math.floor(tile.id / 5) === 0) {
    return {
      ...tile,
      marked: true,
    };
  }
  return tile;
});

export const mockWonBoardDiagonal = baseBoard.map((tile) => {
  // board win by backward slash
  if (tile.id % 6 === 0) {
    return {
      ...tile,
      marked: true,
    };
  }
  return tile;
});
