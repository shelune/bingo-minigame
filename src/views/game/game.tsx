import React, { useState } from "react";
import Board, { Tile } from "../../components/board/board";
import Profile, { PlayerData } from "../../components/profile/profile";
import { WinCondition } from "./data";

import css from "./game.module.scss";

type Props = {
  username: string;
  tiles: Tile[];
  onMark: (tiles: Tile[], id: number) => void;
  onAdvanceTurn: () => void;
  onReset: () => void;
  winCondition: WinCondition;
  picked: number[];
  elapsedTime: number;
  profile: PlayerData | null;
  onDeleteProfile: () => void;
};

export function GameView({
  username,
  tiles,
  onMark,
  onAdvanceTurn,
  onReset,
  winCondition,
  picked,
  elapsedTime,
  profile,
  onDeleteProfile,
}: Props) {
  const [profileIsOpen, setProfileIsOpen] = useState(false);
  const pickedValue = picked.map((number) =>
    number === 12 ? "Free" : tiles[number].value
  );
  return (
    <div className={css.content}>
      <Profile
        isOpen={profileIsOpen}
        setOpen={() => setProfileIsOpen(false)}
        playerData={profile}
        deleteProfile={onDeleteProfile}
      />
      <div className={css.header}>
        <div className={css.leftSide}>A game of Bingo</div>
        <div className={css.rightSide}>
          <div
            className={css.profile}
            role="button"
            onClick={() => setProfileIsOpen(true)}
          >
            {username}
          </div>
        </div>
      </div>
      <div className={css.gameView}>
        <Board tiles={tiles} onMark={onMark} />
        <div className={css.announcement}>
          <div>Tiles picked: {pickedValue.join(", ")}</div>
          {!!winCondition ? (
            <div className={css.victory}>
              ðŸŽ‰ You won by matching {winCondition} after {picked.length} turns.
              The game lasted for {elapsedTime} seconds ðŸŽ‰
            </div>
          ) : null}
        </div>
        <div className={css.interactions}>
          <button
            disabled={!!winCondition}
            type="button"
            className={css.button}
            onClick={onAdvanceTurn}
          >
            Get Number
          </button>
          <button type="button" className={css.button} onClick={onReset}>
            Reset Board
          </button>
        </div>
      </div>
    </div>
  );
}
