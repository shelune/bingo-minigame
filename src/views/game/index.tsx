import React from "react";
import { useData } from "./data";
import { GameView as MainComponent } from "./game";

type Props = {
  username: string;
};

export const GameView = ({ username }: Props) => {
  const props = useData();
  // eslint-disable-next-line react/jsx-props-no-spreading
  return <MainComponent username={username} {...props} />;
};
