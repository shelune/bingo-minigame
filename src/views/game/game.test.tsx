import React from "react";
import { screen, render } from "@testing-library/react";

import { GameView } from "./game";
import { mockBareBoard } from "../../views/game/utils.mock";

describe("Board", () => {
  it("should render", () => {
    const { container } = render(
      <GameView
        username="foobar"
        tiles={mockBareBoard}
        onMark={jest.fn()}
        onAdvanceTurn={jest.fn()}
        onReset={jest.fn()}
        picked={[]}
        winCondition=""
        elapsedTime={0}
        profile={null}
        onDeleteProfile={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("should render announcement on a won board & disable Get Number", () => {
    // not following actual data correctness
    // just want to test the announcement markup / visual
    // since otherwise we have to check if the board is already won at start
    // which is a bit illogical for this kind of game
    render(
      <GameView
        username="foobar"
        tiles={mockBareBoard}
        onMark={jest.fn()}
        onAdvanceTurn={jest.fn()}
        onReset={jest.fn()}
        picked={[0, 1, 2, 3, 4]}
        winCondition="row"
        elapsedTime={5}
        profile={null}
        onDeleteProfile={jest.fn()}
      />
    );
    expect(screen.getByTestId("announcement")).toMatchSnapshot();
    expect(screen.getByText("Get Number")).toHaveAttribute("disabled");
  });
});
