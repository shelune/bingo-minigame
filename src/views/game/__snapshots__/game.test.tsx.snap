// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="content"
      data-testid="profile-drawer"
    >
      <div
        class="profileHeader"
      >
        <div
          class="closeButton"
          role="button"
        >
          &gt; Close
        </div>
        <h2>
          Profile
        </h2>
      </div>
      <div
        class="profileError"
      >
        No record of this player yet. Please play some games first!
      </div>
    </div>
    <div
      class="header"
    >
      <div
        class="leftSide"
      >
        A game of Bingo
      </div>
      <div
        class="rightSide"
      >
        <div
          class="profile"
          role="button"
        >
          foobar
        </div>
      </div>
    </div>
    <div
      class="gameView"
    >
      <div
        class="content"
      >
        <h1
          class="title"
        >
          <div>
            <span>
              B
            </span>
          </div>
          <div>
            <span>
              I
            </span>
          </div>
          <div>
            <span>
              N
            </span>
          </div>
          <div>
            <span>
              G
            </span>
          </div>
          <div>
            <span>
              O
            </span>
          </div>
        </h1>
        <div
          class="board"
        >
          <div
            class="tile"
            data-testid="tile"
          >
            2
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            16
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            39
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            55
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            61
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            9
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            25
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            32
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            53
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            67
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            10
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            19
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            Free
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            59
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            63
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            13
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            27
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            43
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            47
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            64
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            7
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            18
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            38
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            58
          </div>
          <div
            class="tile"
            data-testid="tile"
          >
            71
          </div>
        </div>
      </div>
      <div
        class="announcement"
        data-testid="announcement"
      >
        <div>
          Tiles picked: 
        </div>
      </div>
      <div
        class="interactions"
      >
        <button
          class="button"
          type="button"
        >
          Get Number
        </button>
        <button
          class="button ghostButton"
          type="button"
        >
          Reset Board
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Board should render announcement on a won board 1`] = `
<div
  class="announcement"
  data-testid="announcement"
>
  <div>
    Tiles picked: 
    2, 16, 39, 55, 61
  </div>
  <div
    class="victory"
  >
    ðŸŽ‰ You won by matching 
    row
     after 
    5
     turns. The game lasted for 
    0
     seconds ðŸŽ‰
  </div>
</div>
`;
